### PR TITLE
Sendrecv overhaul: async sniffing & major cleanup

### PR DESCRIPTION
**Pull request by gpotter2**
This PR:
- ✔️ ~~**REQUIRES https://github.com/secdev/scapy/pull/2035**~~ 
- **contains an alternative fix for #1199**, that binds the `sent_time` in `__iter__` rather than requiring extra memory space to store the timestamps. This brings a **small performance boost, see below**.
- **Implements `AsyncSniff`**, removes tons of duplications, and allow us to use `Sessions` everywhere
    - `sniff` uses `AsyncSniff`
    - `sndrcv` uses `AsyncSniff`
    - `sndrcvflood` uses `sndrcv`.
- use classes to handle sniffing and sndrcv: they allow greater flexibility when sharing properties than the previous method.
- drops `Event` usage: not required anymore if we use a class.
- removes `process` from sndrcv: no one uses it, and it's extremly messy. (My bad). We could reimplement it much easily now that we use `AsyncSniffer` within `sndrcv`
- fixes *heavy* memory leaks in `scapy/automaton.py`
- Add docs about `AsyncSniffer`
- ❕ **Change of behavior: packets that are `None` are ignored on all platforms (previously only Windows/BPF), instead of closing the socket (and using the very unclear `TimeoutElapsed`). `EOFError` will now be the default exception to close a socket (this behavior is only used in `*PcapReader`)**
- **fixed BSD bug: `conf.use_bpf` would always enable itself even if `conf.use_pcap` was True. (Missing parenthesis)**
- migrates `scapy.arch.bpf.supersocket` sockets to use the `recv_raw` mechanic

It's a bit harder to implement than it looks, because it must be able to get out of a frozen `select()`, on sockets that don't get any more packets.

#### Perfs:

Using the benchmarking of https://github.com/secdev/scapy/pull/1259

**Note: the following stats should only be read relatively speaking. The test is performed using a short packet (`Ether()/IP()`) on loopback. It isn't intended to match a real usage by any mean.**

|            | secdev/master | gpotter2/perfsndrcv |
|------------|---------------|---------------------|
| Python 2.7 | ~ 5700/s      | ~ 5900/s            |
| Python 3.7 | ~ 6800/s      | ~ 7300/s            |

#### Notes:

`async_select_unrequired=False` means that the `select` function is blocking, therefore a control socket (fake) should be passed. It should be set to `True` otherwise (Windows/OSX/Pcapdnet)... We could potentially rename it `is_blocking_socket=False` if it's clearer (but I only thought of that later 😄)

#### Hopes for the future:

Thanks to the use of *sniff within the sndrcv suite, we have less code to maintain + it will allow us to enable:
- `Sessions` in `sndrcv*`: match a chunked HTTP response to its answer ?
- multi-thread dissection (I have great hopes for this. I plan to use [`multiprocessing.dummy`](https://docs.python.org/3/library/multiprocessing.html#module-multiprocessing.dummy) and try to paralellize packet dissections in pools)

#### Related:

fixes https://github.com/secdev/scapy/pull/1523
fixes https://github.com/secdev/scapy/issues/1505
fixes https://github.com/secdev/scapy/pull/1937

### CODE REVIEW

File: scapy/fields.py [hunk 1]
```diff
@@ -1967,11 +1967,7 @@ def _fixup_val(self, x):
 
         """
         if isinstance(x, (list, tuple)):
-            return type(x)(
-                v if v is None or isinstance(v, FlagValue)
-                else FlagValue(v, self.names)
-                for v in x
-            )
+            return FlagValue(list(x), self.names)
```
Comment (gpotter2) [inline scapy/fields.py]: This was super glitchy:
**before (?!?!)**
```python
>>> EIR_Hdr() / EIR_Flags(flags=["general_disc_mode", "br_edr_not_supported"])
<EIR_Hdr  type=flags |<EIR_Flags  flags=['general_disc_mode', 'br_edr_not_supported'] |>>
```
**after**
```python
>>> EIR_Hdr() / EIR_Flags(flags=["general_disc_mode", "br_edr_not_supported"])
<EIR_Hdr  type=flags |<EIR_Flags  flags=general_disc_mode+br_edr_not_supported |>>
```

File: scapy/contrib/ibeacon.uts [hunk 2]
```diff
@@ -42,7 +42,7 @@ assert d == d2
 
 = iBeacon (encode LE Set Advertising Data)
 
-d = hex_bytes('1E0201021AFF4C000215FB0B57A2822844CD913A94A122BA120600010002D100')
+d = hex_bytes('1E0201061AFF4C000215FB0B57A2822844CD913A94A122BA120600010002D100')
```
Comment (gpotter2) [inline scapy/contrib/ibeacon.uts]: This was triggered by the `FlagsField` bug

File: scapy/packet.py [hunk 3]
```diff
@@ -838,9 +847,12 @@ def guess_payload_class(self, payload):
...
             for fval, cls in t.payload_guess:
-                if all(hasattr(self, k) and v == self.getfieldval(k)
-                       for k, v in six.iteritems(fval)):
-                    return cls
+                try:
+                    if all(v == self.getfieldval(k)
+                           for k, v in six.iteritems(fval)):
+                        return cls
+                except AttributeError:
+                    pass
```
Comment (gpotter2) [inline scapy/packet.py]: This is actually quite a big improvement: in fact `hasattr` calls `__getattr__` under the hood, which calls `getfieldval` => we actually did `getfieldval` twice each time.

File: scapy/packet.py [hunk 4]
```diff
@@ -897,46 +911,42 @@ def loop(todo, done, self=self):
...
+                if self.fields == done and payloads.__iterlen__() == 1:
+                    # Only one payload + same fields
+                    yield self
+                    return
                 for payl in payloads:
-                    done2 = done.copy()
-                    for k in done2:
-                        if isinstance(done2[k], VolatileValue):
-                            done2[k] = done2[k]._fix()
-                    pkt = self.clone_with(payload=payl, **done2)
```
Comment (gpotter2) [inline scapy/packet.py]: We don't need this anymore since we allow `VolatileValue` in Packet

File: scapy/packet.py [hunk 5]
```diff
@@ -897,46 +911,42 @@ def loop(todo, done, self=self):
...
+        fields = list(self.fields)
         length = 1
         for field in fields:
             fld, val = self.getfield_and_val(field)
             if hasattr(val, "__iterlen__"):
                 length *= val.__iterlen__()
-            elif isinstance(val, tuple) and len(val) == 2 and all(hasattr(z, "__int__") for z in val):  # noqa: E501
-                length *= (val[1] - val[0])
+            elif isinstance(val, tuple) and len(val) == 2 and all(isinstance(z, int) for z in val):  # noqa: E501
+                length *= (val[1] - val[0] + 1)
```
Comment (gpotter2) [inline scapy/packet.py]: For some reason I missed that: `(0, 5)` makes 6 values... -_-

File: test/regression.uts [hunk 6]
```diff
@@ -7038,8 +7054,6 @@ assert isinstance(pkt[LLTDAttributeEOP].payload, NoPayload)
 = Packet build / dissection
 pkt = Ether(raw(Ether(dst=ETHER_BROADCAST, src=RandMAC()) / LLTD(tos=0, function=0)))
 assert LLTD in pkt
-assert pkt.dst == pkt.real_dst
-assert pkt.src == pkt.real_src
```
Comment (gpotter2) [inline test/regression.uts]: This makes literally no sense. I wonder how it worked previously :/
There's absolutely no method in `LLTD` that ensures that the `real_dst` would match the parent one

File: test/regression.uts [hunk 7]
```diff
@@ -10653,15 +10667,6 @@ assert not any(getattr(p1.flags | p2.flags, f) for f in 'FRPECN')
...
 assert TCP(flags="SA").flags | TCP(flags="S").flags == TCP(flags="SA").flags
 
-= Using tuples and lists as flag values
-~ IP TCP
-
-plist = PacketList(list(IP()/TCP(flags=(0, 2**9 - 1))))
-assert [p[TCP].flags for p in plist] == [x for x in range(512)]
-
-plist = PacketList(list(IP()/TCP(flags=["S", "SA", "A"])))
-assert [p[TCP].flags for p in plist] == [2, 18, 16]
```
Comment (gpotter2) [inline test/regression.uts]: This also made very little sense: it abuses the bugs in `__iter__` to generate a list of packets, whereas it should only have been a flag field.

**Comment by gpotter2** [comment]
PR is currently failing with PCAP/DNET tests. I'm not sure why yet

**Comment by gpotter2** [comment]
So I've been working on asynchronous sniffing for a while now, PR isn't finished (fails on OSX/pcapdnet), but at least it passes on Windows and Linux (I have access to debug machines)

@polybassa I'm pretty sure this could be used in the `isotp` implementation. However, I'm having troubles getting it to work with ISOTP sockets: https://travis-ci.com/secdev/scapy/jobs/203122929

If you could have a look it would be amazing, but the PR is quite hard to understand, so it's up to you.

**Comment by polybassa** [comment]
Hi, I’ll check that PR. Unfortunately I don’t have time for it, today. But tomorrow I’ll look into it.

**Comment by polybassa** [comment]
Hi, I've looked into your commit and tested it manually. None of my unit tests for ISOTP fails at my local machine. AsyncSniffer works fine with ISOTPNativeSocket and ISTOPSoftSocket if the socket is provided as parameter `opened_socket`. Do you have any specific place, where ISOTPSocket fail?

**Comment by gpotter2** [comment]
Thanks for the feedback I've actually updated the `StreamSocket` a bit since the commit. It might have been related to that.

It *seems* to have been fixed, sorry for wasting your time. I'll see if it remains consistent

**Comment by polybassa** [comment]
No problem. I'm happy to help you.

**Comment by gpotter2** [comment]
PR Status update: fixes for Pcapdnet. Works on Linux/Windows (with/out) pcapdnet.
Still doesn't pass on OSX, but I'm blind testing it so it might take some time...

**Comment by gpotter2** [comment]
I managed to circle the issue. It's definetly not what I expected :/
```python
if timeout is not None:
    def _timeout():
        time.sleep(timeout)
        self.stop_cb()
    self._timeout = Thread(
        target=_timeout
    )
    self._timeout.setDaemon(True)
```
freezes the main thread when running :/ Even if it's in another thread. It looks like some BSD/Python glitch

**Comment by gpotter2** [comment]
@polybassa I finally managed to fix BSD*. The last commit fails because of `isotp.uts` tests (see the Travis build: https://travis-ci.com/secdev/scapy/builds/114015453) :/

The PR as right nearly is in its final state, the only thing left is to fix that. It works all right if I disable those.
I'll try to figure it out, but you're much more familiar with this part than I am. I would be very grateful if you could have a (another?) look.

Also, note that this PR will probably allow to rewrite `CANReceiverThread` and `ISOTPSniffer`, in a prettier way.

**Comment by polybassa** [comment]
@gpotter2 
The tests that are failing, are tests with a mocked socket. This means, the problem is not related to vcan or to CANSockets. Either the implementation of the MockSocket has a bug or the implementation of ISOTPSoftSocket is the problem. Maybe @epozzobon can help us here.

**Comment by gpotter2** [comment]
@polybassa You were very right ! Thanks a lot, I had missed the `CANMockSocket` :smile:

I've updated it using the new mechanic in https://github.com/secdev/scapy/commit/be21a504f62233c67b74a29fe209a791b2b9a3d2, and it appears to have solved the issue !

**Comment by gpotter2** [comment]
@guedou @p-l- This PR is ready to be reviewed.
It's quite big, and it did expand a bit more than I expected. It's still a pretty cool feature

File: .travis.yml [hunk 8]
```diff
@@ -134,6 +134,8 @@ matrix:
           env:
             - TOXENV=linux_warnings
 
-install: bash .travis/install.sh
+install:
+  - bash .travis/install.sh
+  - python -c "from scapy.all import conf; print(repr(conf))"
```
Comment (guedou) [inline .travis.yml]: Is this needed?

File: scapy/arch/bpf/supersocket.py [hunk 9]
```diff
@@ -208,11 +207,15 @@ def close(self):
 
     def send(self, x):
         """Dummy send method"""
-        raise Exception("Can't send anything with %s" % self.__name__)
+        raise Exception(
```
Comment (guedou) [inline scapy/arch/bpf/supersocket.py]: Unecessayr change.

File: scapy/arch/bpf/supersocket.py [hunk 10]
```diff
@@ -208,11 +207,15 @@ def close(self):
...
+            "Can't send anything with %s" % self.__class__.__name__
+        )
 
-    def recv(self, x=BPF_BUFFER_LENGTH):
+    def recv_raw(self, x=BPF_BUFFER_LENGTH):
         """Dummy recv method"""
-        raise Exception("Can't recv anything with %s" % self.__name__)
+        raise Exception(
```
Comment (guedou) [inline scapy/arch/bpf/supersocket.py]: unnecessary change.

File: scapy/contrib/cansocket_native.py [hunk 11]
```diff
@@ -28,6 +28,7 @@
 
 class CANSocket(SuperSocket):
     desc = "read/write packets at a given CAN interface using PF_CAN sockets"
+    async_select_unrequired = True
```
Comment (guedou) [inline scapy/contrib/cansocket_native.py]: That's a strange name =)

File: scapy/contrib/isotp.uts [hunk 12]
```diff
@@ -737,9 +738,9 @@ with ISOTPSoftSocket(cans, sid=0x641, did=0x241) as s:
             raise ex
     thread = threading.Thread(target=sender, name="sender")
     thread.start()
-    ready.wait()
+    ready.wait(15)
```
Comment (guedou) [inline scapy/contrib/isotp.uts]: Is there a need to configure this value?

File: scapy/arch/bpf/supersocket.py (hunk 10 above)
Comment (gpotter2) [reply scapy/arch/bpf/supersocket.py]: It's not.
`.__name__` doesn't exist on an instance.
However, `.__class__.__name__` does

File: scapy/arch/bpf/supersocket.py (hunk 9 above)
Comment (gpotter2) [reply scapy/arch/bpf/supersocket.py]: Same

File: .travis.yml (hunk 8 above)
Comment (gpotter2) [reply .travis.yml]: It's not shown by default (requires to be clicked on), and will help to debug a lot of things.
I'd say it's fine to leave it

File: scapy/contrib/isotp.uts (hunk 12 above)
Comment (gpotter2) [reply scapy/contrib/isotp.uts]: I don't think so. This is just to avoid Travis freeze, and rather have the error displayed.

File: scapy/contrib/cansocket_native.py (hunk 11 above)
Comment (gpotter2) [reply scapy/contrib/cansocket_native.py]: Renamed `nonblocking_socket`, should be clearer :smile:

**Comment by porceCodes** [comment]
Document:
https://scapy.readthedocs.io/en/latest/usage.html#asynchronous-sniffing
